### PR TITLE
Support iproute2 bpf_map_def layout

### PR DESCRIPTION
**Pull request by RafaelOrtizRC**
Addresses #10 

Please note, while this passes tests I haven't had a chance to actually try it with an older/new map definitions. If someone has an old program lying around, please test it before merging.

It also breaks the `id` and `pinning` fields, replacing them with the members found [here](https://elixir.bootlin.com/linux/v5.12.12/source/include/uapi/linux/bpf.h#L491).

### CODE REVIEW

**Comment by alessandrod** [comment]
`bpf_map_def` is not the same as the args passed to `BPF_MAP_CREATE`. 

There are two versions out in the wild, one used by libbpf: https://github.com/libbpf/libbpf/blob/12eb2666d9e97d03a10347edb12112b3f638b989/src/libbpf.h#L400 and one used by iproute2: https://github.com/eworm-de/iproute2/blob/1fdea280517b9951e43bc4544210024d0346b27e/include/bpf_elf.h#L32

`id` and `pinning` are required to load programs originally written to be loaded with `tc` and friends from iproute2.

**Comment by RafaelOrtizRC** [comment]
> `bpf_map_def` is not the same as the args passed to `BPF_MAP_CREATE`.

Whoops, I misunderstood how the values were being passed around. It's using the correct struct now.

File: aya/src/bpf.rs [hunk 1]
```diff
@@ -48,15 +48,17 @@ unsafe_impl_pod!(i8, u8, i16, u16, i32, u32, i64, u64);
...
+    // minimum features required by old BPF programs
     pub(crate) map_type: u32,
     pub(crate) key_size: u32,
     pub(crate) value_size: u32,
     pub(crate) max_entries: u32,
     pub(crate) map_flags: u32,
-    pub(crate) id: u32,
-    pub(crate) pinning: u32,
+    // optional features
+    pub(crate) id: u32, // id
```
Comment (alessandrod) [inline aya/src/bpf.rs:60]:
```suggestion
    pub(crate) id: u32,
```

File: aya/src/bpf.rs [hunk 2]
```diff
@@ -48,15 +48,17 @@ unsafe_impl_pod!(i8, u8, i16, u16, i32, u32, i64, u64);
...
+    pub(crate) pinning: u32,    // pinning
```
Comment (alessandrod) [inline aya/src/bpf.rs:61]:
```suggestion
    pub(crate) pinning: u32,
```

File: aya/src/obj/mod.rs [hunk 3]
```diff
@@ -510,13 +509,24 @@ fn parse_map(section: &Section, name: &str) -> Result<Map, ParseError> {
 }
 
 fn parse_map_def(name: &str, data: &[u8]) -> Result<bpf_map_def, ParseError> {
-    if mem::size_of::<bpf_map_def>() > data.len() {
+    if data.len() > mem::size_of::<bpf_map_def>() {
```
Comment (alessandrod) [inline aya/src/obj/mod.rs:512]: We also need to check that it's at least as big as up to `map_flags`

File: aya/src/obj/mod.rs [hunk 4]
```diff
@@ -510,13 +509,24 @@ fn parse_map(section: &Section, name: &str) -> Result<Map, ParseError> {
...
         return Err(ParseError::InvalidMapDefinition {
             name: name.to_owned(),
         });
     }
 
-    Ok(unsafe { ptr::read_unaligned(data.as_ptr() as *const bpf_map_def) })
+    if data.len() <  mem::size_of::<bpf_map_def>() {
+        let mut map_def = bpf_map_def { ..Default::default() };
```
Comment (alessandrod) [inline aya/src/obj/mod.rs:519]:
```suggestion
        let mut map_def = bpf_map_def::default();
```

File: aya/src/obj/mod.rs [hunk 5]
```diff
@@ -510,13 +509,24 @@ fn parse_map(section: &Section, name: &str) -> Result<Map, ParseError> {
...
+        unsafe {
+            ptr::copy(data.as_ptr() as *const bpf_map_def, &mut map_def as *mut bpf_map_def, 1);
+            // id and pinning will be garbage data
```
Comment (alessandrod) [inline aya/src/obj/mod.rs:522]: This is UB, see https://doc.rust-lang.org/stable/std/ptr/fn.copy.html#safety

You need to cast both to bytes and copy bytes. After copying bytes, resetting id and pinning won't be necessary.

File: aya/src/obj/mod.rs [hunk 6]
```diff
@@ -657,10 +667,19 @@ mod tests {
 
     #[test]
     fn test_parse_map_def() {
+        #![allow(unused_variables)] // map_def is used by the assertion
         assert!(matches!(
-            parse_map_def("foo", &[]),
+            parse_map_def("foo", &[0u8; std::mem::size_of::<bpf_map_def>() + 10]),
```
Comment (alessandrod) [inline aya/src/obj/mod.rs:672]: The old test is still valid, please revert this change

File: aya/src/obj/mod.rs [hunk 7]
```diff
@@ -670,32 +689,45 @@ mod tests {
...
-                id: 0,
-                pinning: 0
-            })
+            Ok(map_def)
+        ));
+
+        assert!(matches!(
+            parse_map_def(
+                "foo",
+                &[0u8, 0u8, 0u8, 1u8, 0u8, 0u8, 0u8, 2u8, 0u8, 0u8, 0u8, 4u8, 0u8, 0u8, 0u8, 5u8]
```
Comment (alessandrod) [inline aya/src/obj/mod.rs:701]: use `bytes_of` here and index with a shorter range

File: aya/src/obj/mod.rs [hunk 8]
```diff
@@ -670,32 +689,45 @@ mod tests {
...
+            ),
+            Ok(map_def)
         ));
+        let map = parse_map_def(
+            "foo",
+            &[0u8, 0u8, 0u8, 1u8, 0u8, 0u8, 0u8, 2u8, 0u8, 0u8, 0u8, 4u8, 0u8, 0u8, 0u8, 5u8]
```
Comment (alessandrod) [inline aya/src/obj/mod.rs:707]: same, use `bytes_of`  please

File: aya/src/obj/mod.rs [hunk 9]
```diff
@@ -670,32 +689,45 @@ mod tests {
...
+        ).unwrap();
+        assert!(map.id == 0 && map.pinning == 0)
     }
 
     #[test]
     fn test_parse_map_error() {
         assert!(matches!(
-            parse_map(&fake_section("maps/foo", &[]), "foo"),
+            parse_map(&fake_section("maps/foo", &[0u8; std::mem::size_of::<bpf_map_def>() + 10]), "foo"),
             Err(ParseError::InvalidMapDefinition { .. })
```
Comment (alessandrod) [inline aya/src/obj/mod.rs:716]: the old test was correct, please revert

File: aya/src/obj/mod.rs [hunk 10]
```diff
@@ -706,32 +738,34 @@ mod tests {
                         value_size: 3,
                         max_entries: 4,
                         map_flags: 5,
-                        id: 0,
```
Comment (alessandrod) [inline aya/src/obj/mod.rs:709]: I think you wanted to replace this with `def` too or why bind `def` at all?

File: aya/src/obj/mod.rs [hunk 11]
```diff
@@ -706,32 +738,34 @@ mod tests {
...
-                },
+                def,
                 data
             }) if name == "foo" && data.is_empty()
         ))
     }
 
     #[test]
     fn test_parse_map_data() {
+        #![allow(unused_variables)] // def is used by the assertion
```
Comment (alessandrod) [inline aya/src/obj/mod.rs:757]: I don't think this test should be changed? Other than the Default::default() bit that is

**Review by alessandrod — CHANGES_REQUESTED** [review]
Thanks for doing this! See my comments

**Comment by alessandrod** [comment]
Oh also please make sure to `cargo fmt` your changes

File: aya/src/obj/mod.rs (hunk 10 above)
Comment (RafaelOrtizRC) [reply aya/src/obj/mod.rs:709]: rustc doesn't like it when `..Default::default()` is called in `Ok()`. 

```
error: expected `}`, found `Default`
   --> aya/src/obj/mod.rs:709:19
    |
703 |             Ok(bpf_map_def {
    |                ----------- while parsing the fields for this pattern
...
709 |                 ..Default::default()
    |                   ^^^^^^^ expected `}`
```

Is there an alternative?

File: aya/src/obj/mod.rs (hunk 11 above)
Comment (RafaelOrtizRC) [reply aya/src/obj/mod.rs:757]: I was getting errors due to `_map_type` and `value_size` being undefined.

File: aya/src/obj/mod.rs (hunk 3 above)
Comment (RafaelOrtizRC) [reply aya/src/obj/mod.rs:512]: I added the check and put back the tests for a too-small definition. Now it should check for too large and too small.

File: aya/src/obj/mod.rs (hunk 5 above)
Comment (RafaelOrtizRC) [reply aya/src/obj/mod.rs]: I was trying to avoid that byte copy loop but you're right - no guarantees about what will happen when copying past the end of `data`.

See 1404d12a7e29ab02225bc13fd36794f6c6206db1

File: aya/src/obj/mod.rs (hunk 6 above)
Comment (RafaelOrtizRC) [reply aya/src/obj/mod.rs:681]: Reverted in 1404d12a7e29ab02225bc13fd36794f6c6206db1

File: aya/src/obj/mod.rs (hunk 7 above)
Comment (RafaelOrtizRC) [reply aya/src/obj/mod.rs]: Fixed in 1404d12a7e29ab02225bc13fd36794f6c6206db1

File: aya/src/obj/mod.rs (hunk 8 above)
Comment (RafaelOrtizRC) [reply aya/src/obj/mod.rs]: Fixed in 1404d12a7e29ab02225bc13fd36794f6c6206db1

File: aya/src/obj/mod.rs (hunk 9 above)
Comment (RafaelOrtizRC) [reply aya/src/obj/mod.rs:747]: Reverted in 1404d12a7e29ab02225bc13fd36794f6c6206db1

File: aya/src/obj/mod.rs [hunk 12]
```diff
@@ -510,13 +511,27 @@ fn parse_map(section: &Section, name: &str) -> Result<Map, ParseError> {
...
         });
     }
 
-    Ok(unsafe { ptr::read_unaligned(data.as_ptr() as *const bpf_map_def) })
+    if data.len() < mem::size_of::<bpf_map_def>() {
+        let mut map_def = bpf_map_def::default();
+        unsafe {
+            let mut data_ptr = data.as_ptr();
+            let mut map_def_ptr = &mut map_def as *mut bpf_map_def as *mut u8;
+            for _ in 0..=data.len() {
```
Comment (alessandrod) [inline aya/src/obj/mod.rs:525]: Instead of looping, please convert to byte slices and use `copy_from_slice`

File: aya/src/obj/mod.rs [hunk 13]
```diff
@@ -657,10 +672,23 @@ mod tests {
 
     #[test]
     fn test_parse_map_def() {
+        #![allow(unused_variables)] // map_def is used by the assertion
```
Comment (alessandrod) [inline aya/src/obj/mod.rs:675]: the compiler is giving you a warning because using `matches!` map_def is not being used, it's being re-bound in the pattern.

If you want to factor out `map_def` - which looks good to me - you need to change the test to use `assert_eq!` instead of `matches!`. Eg:
```suggestion
assert_eq!(parse_map_def("foo", bytes_of(&def)).unwrap(), def)
```

For that to work, you'll have to make `bpf_map_def`  derive `PartialEq`

File: aya/src/obj/mod.rs [hunk 14]
```diff
@@ -657,10 +672,23 @@ mod tests {
...
         assert!(matches!(
             parse_map_def("foo", &[]),
             Err(ParseError::InvalidMapDefinition { .. })
         ));
+        assert!(matches!(
+            parse_map_def("foo", &[0u8; std::mem::size_of::<bpf_map_def>() + 10]),
```
Comment (alessandrod) [inline aya/src/obj/mod.rs:681]:
```suggestion
            parse_map_def("foo", &[0u8; std::mem::size_of::<bpf_map_def>() + 1]),
```

File: aya/src/obj/mod.rs [hunk 15]
```diff
@@ -670,32 +698,67 @@ mod tests {
...
 
     #[test]
     fn test_parse_map_error() {
         assert!(matches!(
             parse_map(&fake_section("maps/foo", &[]), "foo"),
             Err(ParseError::InvalidMapDefinition { .. })
+        ));
+        assert!(matches!(
+            parse_map(
+                &fake_section("maps/foo", &[0u8; std::mem::size_of::<bpf_map_def>() + 10]),
```
Comment (alessandrod) [inline aya/src/obj/mod.rs:744]:
```suggestion
                &fake_section("maps/foo", &[0u8; std::mem::size_of::<bpf_map_def>() + 1]),
```

File: aya/src/obj/mod.rs (hunk 10 above)
Comment (alessandrod) [reply aya/src/obj/mod.rs:709]: yes, switch to `assert_eq` see the comment above

**Review by alessandrod — CHANGES_REQUESTED** [review]

File: aya/src/obj/mod.rs (hunk 15 above)
Comment (RafaelOrtizRC) [reply aya/src/obj/mod.rs]: Fixed in 86b639b

File: aya/src/obj/mod.rs (hunk 14 above)
Comment (RafaelOrtizRC) [reply aya/src/obj/mod.rs]: Fixed in 86b639b

File: aya/src/obj/mod.rs (hunk 13 above)
Comment (RafaelOrtizRC) [reply aya/src/obj/mod.rs]: Fixed in 86b639b

File: aya/src/obj/mod.rs (hunk 12 above)
Comment (RafaelOrtizRC) [reply aya/src/obj/mod.rs]: Fixed in 86b639b

File: aya/src/obj/mod.rs (hunk 11 above)
Comment (RafaelOrtizRC) [reply aya/src/obj/mod.rs]: Fixed in 86b639b

File: aya/src/obj/mod.rs (hunk 10 above)
Comment (RafaelOrtizRC) [reply aya/src/obj/mod.rs]: Fixed in 86b639b
